File: backend/F1Fantasy/Controllers/DriverStandingController.cs
using F1Fantasy.Models;
using F1Fantasy.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.RateLimiting;

namespace F1Fantasy.Controllers;

[Authorize]
[ApiController]
[Route("api/[controller]")]
[EnableRateLimiting("read")]
public class DriverStandingController : ControllerBase
{
    private readonly DriverStandingService _service;
    private readonly ILogger<DriverStandingController> _logger;

    public DriverStandingController(DriverStandingService service, ILogger<DriverStandingController> logger)
    {
        _service = service;
        _logger = logger;
    }

    [HttpGet("season/{season}")]
    public async Task<ActionResult<StandingsList>> GetDriverStandingsBySeason(string season)
    {
        _logger.LogInformation("GET request for driver standings: season {Season}", season);
        
        try
        {
            var standings = await _service.GetDriverStandingsBySeasonAsync(season);
            
            if (standings == null)
            {
                _logger.LogWarning("No driver standings found for season {Season}", season);
                return NotFound(new ErrorResponse
                {
                    Message = $"No driver standings found for season {season}",
                    StatusCode = 404
                });
            }

            _logger.LogDebug("Returning {Count} driver standings for season {Season}, round {Round}", 
                standings.DriverStandings?.Count ?? 0, season, standings.Round);
            return Ok(standings);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching driver standings for season {Season}", season);
            return StatusCode(500, new ErrorResponse
            {
                Message = "An error occurred while fetching driver standings",
                StatusCode = 500
            });
        }
    }

    [HttpGet("season/{season}/round/{round}")]
    public async Task<ActionResult<StandingsList>> GetDriverStandingsByRound(string season, string round)
    {
        _logger.LogInformation("GET request for driver standings: season {Season}, round {Round}", 
            season, round);
        
        try
        {
            var standings = await _service.GetDriverStandingsByRoundAsync(season, round);
            
            if (standings == null)
            {
                _logger.LogWarning("No driver standings found for season {Season}, round {Round}", season, round);
                return NotFound(new ErrorResponse
                {
                    Message = $"No driver standings found for season {season}, round {round}",
                    StatusCode = 404
                });
            }

            _logger.LogDebug("Returning {Count} driver standings for season {Season}, round {Round}", 
                standings.DriverStandings?.Count ?? 0, season, round);
            return Ok(standings);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching driver standings for season {Season}, round {Round}", 
                season, round);
            return StatusCode(500, new ErrorResponse
            {
                Message = "An error occurred while fetching driver standings",
                StatusCode = 500
            });
        }
    }

    [HttpGet("season/{season}/round/{round}/driver/{driverId}")]
    public async Task<ActionResult<DriverStanding>> GetDriverStandingByDriver(string season, string round, string driverId)
    {
        _logger.LogInformation("GET request for driver standing: season {Season}, round {Round}, driver {DriverId}", 
            season, round, driverId);
        
        try
        {
            var standing = await _service.GetDriverStandingByDriverAsync(season, round, driverId);
            
            if (standing == null)
            {
                _logger.LogWarning("No standing found for driver {DriverId}, season {Season}, round {Round}", 
                    driverId, season, round);
                return NotFound(new ErrorResponse
                {
                    Message = $"No standing found for driver {driverId}, season {season}, round {round}",
                    StatusCode = 404
                });
            }

            _logger.LogDebug("Returning standing for driver {DriverId}: position {Position}", 
                driverId, standing.Position);
            return Ok(standing);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching standing for driver {DriverId}, season {Season}, round {Round}", 
                driverId, season, round);
            return StatusCode(500, new ErrorResponse
            {
                Message = "An error occurred while fetching driver standing",
                StatusCode = 500
            });
        }
    }

    [HttpGet("cached")]
    public async Task<ActionResult<IEnumerable<DriverStanding>>> GetCachedStandings()
    {
        _logger.LogInformation("GET request for all cached driver standings");
        
        try
        {
            var standings = await _service.GetCachedStandingsAsync();
            _logger.LogDebug("Returning {Count} cached driver standings", standings.Count());
            return Ok(standings);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching cached driver standings");
            return StatusCode(500, new ErrorResponse
            {
                Message = "An error occurred while fetching cached driver standings",
                StatusCode = 500
            });
        }
    }
}

File: backendJava/f1fantasy/src/main/java/no/f1fantasy/controller/DriverStandingController.java
package no.f1fantasy.controller;

import java.util.HashMap;
import java.util.Map;

import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import no.f1fantasy.entity.DriverStanding;
import no.f1fantasy.service.DriverStandingService;

@RestController
@RequestMapping("/api/driver-standing")
@PreAuthorize("isAuthenticated()")
@SuppressWarnings("null")
public class DriverStandingController {

    private final DriverStandingService driverStandingService;

    public DriverStandingController(DriverStandingService driverStandingService) {
        this.driverStandingService = driverStandingService;
    }

    @GetMapping("/season/{season}")
    public ResponseEntity<?> getDriverStandingsBySeason(@PathVariable String season) {
        try {
            java.util.List<DriverStanding> standings = driverStandingService.getDriverStandingsBySeason(season);
            if (standings == null || standings.isEmpty()) {
                return ResponseEntity.status(404).body(errorMap("No driver standings found for season " + season));
            }
            return ResponseEntity.ok(standings);
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching driver standings for season " + season));
        }
    }

    @GetMapping("/season/{season}/round/{round}")
    public ResponseEntity<?> getDriverStandingsByRound(@PathVariable String season, @PathVariable String round) {
        try {
            java.util.List<DriverStanding> standings = driverStandingService.getDriverStandingsByRound(season, round);
            if (standings == null || standings.isEmpty()) {
                return ResponseEntity.status(404).body(errorMap("No driver standings found for season " + season + ", round " + round));
            }
            return ResponseEntity.ok(standings);
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching driver standings for season " + season + ", round " + round));
        }
    }

    @GetMapping("/season/{season}/round/{round}/driver/{driverId}")
    public ResponseEntity<?> getDriverStandingByDriver(@PathVariable String season, @PathVariable String round, @PathVariable String driverId) {
        try {
            java.util.Optional<DriverStanding> standing = driverStandingService.getDriverStandingByDriver(season, round, driverId);
            if (!standing.isPresent()) {
                return ResponseEntity.status(404).body(errorMap("No driver standing found for driver " + driverId + " in season " + season + ", round " + round));
            }
            return ResponseEntity.ok(standing.get());
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching driver standing for driver " + driverId));
        }
    }

    private Map<String, String> errorMap(String message) {
        Map<String, String> error = new HashMap<>();
        error.put("error", message);
        return error;
    }
}
